thirdparty/DiskANN/src/pq_flash_index.cpp

```diff
@@ -332,32 +338,48 @@ namespace diskann {
       return;
     }
 
-    std::vector<int64_t> tmp_result_ids_64(sample_num, 0);
-    std::vector<float>    tmp_result_dists(sample_num, 0);
+    int64_t tmp_result_ids_64;
+    float tmp_result_dists;
 
-#pragma omp parallel for schedule(dynamic, 1) num_threads(nthreads)
-    for (_s64 i = 0; i < (int64_t) sample_num; i++) {
-      cached_beam_search(samples + (i * sample_aligned_dim), 1, l_search,
-                         tmp_result_ids_64.data() + (i * 1),
-                         tmp_result_dists.data() + (i * 1), beamwidth);
+    auto id = 0;
+    while (this->search_counter < sample_num && id < sample_num &&
+          !this->semaph.IsWaitting()) {
+      cached_beam_search(samples + (id * sample_aligned_dim), 1, l_search,
+                        &tmp_result_ids_64, &tmp_result_dists, beamwidth);
+      id++;
+    }
+
+    if (this->semaph.IsWaitting()) {
+      this->semaph.Signal();
+      return;
     }
 
     std::sort(this->node_visit_counter.begin(), node_visit_counter.end(),
               [](std::pair<_u32, _u32> &left, std::pair<_u32, _u32> &right) {
                 return left.second > right.second;
               });
+  
+    std::vector<uint32_t> node_list;
     node_list.clear();
     node_list.shrink_to_fit();
     node_list.reserve(num_nodes_to_cache);
     for (_u64 i = 0; i < num_nodes_to_cache; i++) {
       node_list.push_back(this->node_visit_counter[i].first);
     }
-    this->count_visited_nodes = false;
+
+    reinterpret_cast<std::atomic<bool> &>(this->count_visited_nodes)
```

Comment:
this is SUPER dirty, I'd turn `count_visited_nodes` into `std::atomic<bool>`

Reply:
updated